Final version of program

Pavlo-ops opened the pull request:
I have fixed issues after review:
- switched to Flow;
- added Models for every layers;
- fixed delete function.

PavloEngineer committed:
Migrated build to version catalogs

PavloEngineer committed:
Made layouts more flexble

PavloEngineer committed:
Fixed creating random id

PavloEngineer committed:
Changed TextBar to Snackbar

PavloEngineer committed:
Added domain layer

PavloEngineer committed:
Run into bug with Hilt

PavloEngineer committed:
Encountered on bug with Hilt

Pavlo-ops committed:
Fixed bugs after review

Pavlo-ops committed:
Integrated Flow

-switched to Flow into Dao, Repository, use case
-using StateFlow for ViewModel

Pavlo-ops committed:
Ready for review

-refactored code

Pavlo-ops committed:
Update MyNotesFragment.kt

EugeneOrlov204 commented on app/src/main/java/com/example/notebook/domain/use_cases/GetNotesUseCase.kt:
```diff
@@ -1,13 +1,13 @@
 class GetNotesUseCase @Inject constructor (private val noteRepository: NoteRepository) {
 
-    fun invoke(): LiveData<List<Note>> {
+    fun invoke(): Flow<List<Note>> {
```
Please read about operator overloading https://kotlinlang.org/docs/operator-overloading.html#invoke-operator

EugeneOrlov204 commented on app/src/main/java/com/example/notebook/presentation/ui/fragments/my_notes/MyNoteViewModel.kt:
```diff
@@ -17,11 +27,22 @@ class MyNoteViewModel @Inject constructor(
+                _allNotes.value = list.map { NoteUI.toNoteUI(it) }
+            }
+        }
+    }
```
I bet you could simplify this to 

`val allNotes: Flow<List<NoteUI>> = getNotesUseCase()`

EugeneOrlov204 replied:
Also, I wonder if you know the difference between using StateFlow and Flow

EugeneOrlov204 commented on app/src/main/java/com/example/notebook/presentation/ui/fragments/my_notes/MyNotesFragment.kt:
```diff
@@ -40,8 +42,13 @@ SearchView.OnQueryTextListener {
+//        viewModel.allNotes.observe(viewLifecycleOwner) {
+//            notesAdapter.submitList(it)
+//        }
+        lifecycleScope.launch {
+            viewModel.allNotes.collect { notes ->
+                notesAdapter.submitList(notes)
+            }
```
I would rather avoid collecting values this way, I am not sure what the difference is but we use this approach in our prod code: 
```
viewModel.yourFlow.flowWithLifecycle(viewLifecycleOwner.lifecycle)
            .onEach { data ->
                // do something with your data here
            }
            .launchIn(viewLifecycleOwner.lifecycleScope)
```

EugeneOrlov204 reviewed (commented):
Bad commit naming

EugeneOrlov204 reviewed (approved):
Overall everything is good, I think you could merge it yet please review my comments first and in future try to give more sensible names to your commits please

Pavlo-ops committed:
Updated work of Flow with viewModels

-switched StateFlow to Flow for viewModels;
- made safer way to observe data from Flow in fragments;
- modified UseCases`s using operator overloading

Pavlo-ops committed:
Updated work of Flow with viewModels

-switched StateFlow to Flow for viewModels;
- made safer way to observe data from Flow in fragments;
- modified UseCases`s using operator overloading
